### PR TITLE
Abstract BBox and provide general functions

### PR DESCRIPTION
**Pull request by pkuyym**
revoles #2356

### CODE REVIEW

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+          Matrix::create(inMatrix->getData() + i * channels * imgSize,
+                         channels,
+                         imgSize,
+                         false,
+                         useGpu);
+      MatrixPtr outTmp = Matrix::create(
+          outMatrix->getData() + offset, imgSize, channels, false, useGpu);
+      inTmp->transpose(outTmp, false);
+    }
+    return channels * imgSize;
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: Why return a size？ can it be got from the outMatrix?

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+    real bboxSize1 = bbox1.getSize();
+    real bboxSize2 = bbox2.getSize();
+
+    return interSize / (bboxSize1 + bboxSize2 - interSize);
+  }
+}
+
+vector<real> encodeBBoxWithVar(const NormalizedBBox& priorBBox,
+                               const vector<real> priorBBoxVar,
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: passing by reference.

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "DetectionUtil.h"
+
+namespace paddle {
+
+size_t appendWithPermute(const MatrixPtr inMatrix,
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: const Matrix&

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+                         size_t height,
+                         size_t width,
+                         size_t outTotalSize,
+                         size_t outOffset,
+                         size_t batchSize,
+                         MatrixPtr outMatrix,
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: Matrix& outMatrix

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+    }
+    if (maxGTIdx != -1) {
+      (*matchIndices)[i] = maxGTIdx;
+      (*matchOverlaps)[i] = maxOverlap;
+    }
+  }
+}
+
+pair<size_t, size_t> generateMatchIndices(
+    const MatrixPtr priorValue,
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: const Matrix&

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+    const size_t numPriorBBoxes,
+    const MatrixPtr gtValue,
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.cpp]: const Matrix&

File: paddle/gserver/layers/DetectionUtil.h [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+  T getCenterY() const { return (yMin + yMax) / 2; }
+
+  T getSize() const { return getWidth() * getHeight(); }
+
+  T xMin;
+  T yMin;
+  T xMax;
+  T yMax;
+  bool isDifficult;
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.h]: add comments for these variables.

File: paddle/gserver/layers/DetectionUtil.h [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+};
+
+struct NormalizedBBox : BBoxBase<real> {
+  NormalizedBBox() : BBoxBase<real>() {}
+};
+
+enum PermMode { NCHWTONHWC, NHWCTONCHW };
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.h]: Please refer to PaddlePaddle naming conventions for constants.  https://github.com/PaddlePaddle/cpp-primer-digest/pull/1/files#diff-04c6e90faac2675aa89e2176d2eec7d8R264

File: paddle/gserver/layers/DetectionUtil.h [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+
+/**
+ * @brief Compute jaccard overlap between two bboxes.
+ * @param bbox1 The first bbox
+ * @param bbox2 The second bbox
+ */
+real jaccardOverlap(const NormalizedBBox& bbox1, const NormalizedBBox& bbox2);
+
+/**
+ * @brief Compute offset parameters between prior bbox and groundtruth bbox
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionUtil.h]: groundtruth -> ground truth

File: paddle/gserver/layers/DetectionUtil.cpp [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+
+size_t appendWithPermute(const Matrix& inMatrix,
+                         size_t height,
+                         size_t width,
+                         size_t outTotalSize,
+                         size_t outOffset,
+                         size_t batchSize,
+                         Matrix& outMatrix,
+                         PermMode permMode,
+                         bool useGpu) {
```
Comment (pkuyym) [inline paddle/gserver/layers/DetectionUtil.cpp]: Done.

**Pull request merged**